src/sql/src/plan/expr.rs

```diff
@@ -397,6 +397,23 @@ impl AbstractExpr for CoercibleScalarExpr {
     ) -> Self::Type {
         match self {
             CoercibleScalarExpr::Coerced(expr) => Some(expr.typ(outers, inner, params)),
+            CoercibleScalarExpr::LiteralRecord(scalars) => {
+                let mut fields = vec![];
+                for (i, scalar) in enumerate(scalars) {
+                    fields.push((
+                        format!("f{}", i + 1).into(),
+                        scalar.typ(outers, inner, params)?,
```

Comment:
Bleh--so we have an issue with our type system that we aggressively move text-looking `CoercibleScalarExpr`s to `text`, when we should remain more agnostic about their types. You can see this here in PG:

```sql
SELECT (ROW('fuzzy dice')).f1;
```
```
ERROR:  failed to find conversion function from unknown to text
```

However, this PR doesn't make the situation meaningfully worse than it currently is, so fine to punt on a fix (which will be complex and require either keeping `CoercibleScalarExpr`s in `ColumnType` or having a shadow schema of the types somewhere).

I filed #10870 about this